src/Workspaces/Remote/Core/Diagnostics/DiagnosticComputer.cs

```diff
@@ -124,28 +165,109 @@ private static ImmutableArray<DiagnosticAnalyzer> GetAnalyzers(BidirectionalMap<
-            var analyzerMap = pooledMap.Object;
-
-            // this follow what we do in DiagnosticAnalyzerInfoCache.CheckAnalyzerReferenceIdentity
-            foreach (var reference in project.Solution.AnalyzerReferences.Concat(project.AnalyzerReferences))
+            /// <summary>
+            /// Cache of <see cref="CompilationWithAnalyzers"/> and a map from analyzer IDs to <see cref="DiagnosticAnalyzer"/>s
+            /// for all analyzers for each project.
+            /// The <see cref="CompilationWithAnalyzers"/> instance is shared between all the following document analyses modes for the project:
+            ///  1. Span-based analysis for active document (lightbulb)
+            ///  2. Background analysis for active and open documents.
+            ///  
+            /// NOTE: We do not re-use this cache for project analysis as it leads to significant memory increase in the OOP process,
+            /// and CWT does not seem to drop entries until ForceGC happens.
+            /// </summary>
+            private static readonly ConditionalWeakTable<Project, CompilationWithAnalyzersData> s_cache
+                = new ConditionalWeakTable<Project, CompilationWithAnalyzersData>();
+
+            private CompilationWithAnalyzersData(CompilationWithAnalyzers compilationWithAnalyzers, BidirectionalMap<string, DiagnosticAnalyzer> analyzerToIdMap)
+            {
+                CompilationWithAnalyzers = compilationWithAnalyzers;
+                AnalyzerToIdMap = analyzerToIdMap;
+            }
+
+            public BidirectionalMap<string, DiagnosticAnalyzer> AnalyzerToIdMap { get; }
+            public CompilationWithAnalyzers CompilationWithAnalyzers { get; }
+
+            public static async Task<CompilationWithAnalyzersData> GetOrCreateAsync(
+                Project project,
+                bool isDocumentAnalysis,
+                CancellationToken cancellationToken)
             {
-                if (!referenceSet.Add(reference.Id))
+                if (!isDocumentAnalysis)
                 {
-                    // already exist
-                    continue;
+                    // Only use cache for document analysis.
+                    return await CreateAsync(project, cancellationToken).ConfigureAwait(false);
                 }
 
-                analyzerMap.AppendAnalyzerMap(reference.GetAnalyzers(project.Language));
+                if (s_cache.TryGetValue(project, out var data))
+                {
+                    return data;
+                }
+
+                data = await CreateAsync(project, cancellationToken).ConfigureAwait(false);
+                return s_cache.GetValue(project, _ => data);
             }
 
-            // convert regular map to bidirectional map
-            return new BidirectionalMap<string, DiagnosticAnalyzer>(analyzerMap);
+            private static async Task<CompilationWithAnalyzersData> CreateAsync(Project project, CancellationToken cancellationToken)
```

Comment:
I'd move these methods, and `s_cache` to the containing class and left CompilationWithAnalyzersCacheEntry as just a simple type with two fields and lo logic.